internal/util/debug/debug.go

```diff
@@ -97,25 +123,98 @@ func Listen(opts *ListenOpts) (*Handler, error) {
 		}),
 	))
 
+	http.HandleFunc("/debug/archive", promhttp.InstrumentHandlerDuration(
+		archiveDurations,
+		http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+			rw.Header().Set("Content-Type", "application/zip")
+
+			rw.Header().Set(
+				"Content-Disposition",
+				fmt.Sprintf("attachment; filename=FerretDB-debug-%d.zip", time.Now().UnixMilli()),
+			)
+
+			ctx := req.Context()
+			zipWriter := zip.NewWriter(rw)
+
+			defer func() {
+				if err := zipWriter.Close(); err != nil {
+					opts.L.Error("Archive handler failed", zap.Error(err))
+					rw.WriteHeader(http.StatusInternalServerError)
+
+					return
+				}
+			}()
+
+			metricsFile, err := zipWriter.Create("metrics")
+			if err != nil {
+				opts.L.Error("Archive handler failed", zap.Error(err))
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			// we use *http.Request instead of http.Get function to provide the ctx
+			scrapeReq := must.NotFail(http.NewRequestWithContext(
+				ctx,
+				http.MethodGet,
+				fmt.Sprintf("http://%s%s", req.Host, "/debug/metrics"),
+				nil,
+			))
+
+			resp, err := http.DefaultClient.Do(scrapeReq)
+			if err != nil {
+				opts.L.Error("Archive handler failed - metrics failed", zap.Error(err))
+				rw.WriteHeader(http.StatusInternalServerError)
+
+				return
+			}
+
+			_, err = io.Copy(metricsFile, resp.Body)
+
+			_ = resp.Body.Close()
```

Comment:
Error from Close() should be logged right?

Reply:
Not really, it returns error to conform to `Closer` interface. The `net/http` calls `Response.Body.Close` without handling error in multiple places: [1](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/net/http/response.go;l=338), [2](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/net/http/client.go;l=708).

In `net/http` docs there's an example with `resp.Body.Close()` not being checked: https://pkg.go.dev/net/http#pkg-overview.

Also PTAL at discussion here: securego/gosec#925.

Overall I didn't check for error, because in previous approach I've used deferred close. I'd suggest to keep it like that to not introduce another 3-5 lines after each Close to handle error that doesn't occur. It'll be also less confusing for the future if someone might want to use deferred call here.